### PR TITLE
BZ1920404: Create migration controller in MTC upgrade

### PR DESCRIPTION
**Pull request by apinnick**
https://bugzilla.redhat.com/show_bug.cgi?id=1920404

For MTC 1.4.0 (Feb 10)

CP to 4.5, 4.6, 4.7

Preview sections:

https://bz1920404_upgrade-mtc--ocpdocs.netlify.app/openshift-enterprise/latest/migration/migrating_3_4/upgrading-migration-tool-3-4.html

https://bz1920404_upgrade-mtc--ocpdocs.netlify.app/openshift-enterprise/latest/migration/migrating_4_1_4/upgrading-migration-tool-4-1-4.html

https://bz1920404_upgrade-mtc--ocpdocs.netlify.app/openshift-enterprise/latest/migration/migrating_4_2_4/upgrading-migration-tool-4-2-4.html

### CODE REVIEW

File: modules/migration-upgrading-migration-tool-4.adoc [hunk 1]
```diff
@@ -4,10 +4,9 @@
 // * migration/migrating_4_2_4/upgrading-migration-tool-4-2-4.adoc
 
 [id='migration-upgrading-migration-tool-4_{context}']
-= Upgrading the {mtc-full} Operator on an {product-title} 4 cluster
+= Upgrading the {mtc-full} on an {product-title} 4 cluster
 
-ifeval::["{mtc-version}" > "1.3"]
-You can upgrade the {mtc-full} ({mtc-short}) Operator on an {product-title} 4 cluster with the Operator Lifecycle Manager.
+You can upgrade the {mtc-full} ({mtc-short}) on an {product-title} 4 cluster by using the Operator Lifecycle Manager (OLM) to install the {mtc-full} Operator.
```
Comment (XinRedhat) [inline modules/migration-upgrading-migration-tool-4.adoc:9]: @apinnick do we need to clarify the upgrade is from MTC 1.3.x -> 1.4.0? If customer is using 1.2.5, the upgrade guide is not the same.

File: modules/migration-upgrading-migration-tool-4.adoc [hunk 2]
```diff
@@ -21,27 +20,42 @@ The following procedure enables you to change the *Manual* approval option to *A
 . Optional: Edit the *Channel*.
 +
 Updating the subscription deploys the updated {mtc-full} Operator and updates the {mtc-short} components.
-endif::[]
-ifeval::["{mtc-version}" <= "1.3"]
-You can upgrade to {mtc-full} ({mtc-short}) {mtc-version} on an {product-title} 4 cluster by deleting the `MigrationController` custom resource (CR), uninstalling the CAM Operator, and then installing the {mtc-full} Operator.
 
-.Procedure
+. Click *{mtc-full} Operator*.
```
Comment (XinRedhat) [inline modules/migration-upgrading-migration-tool-4.adoc:24]: As the instruction is to describe the MTC upgrade from 1.3.x -> 1.4.0, step #5,6,7 should be removed.

File: modules/migration-upgrading-migration-tool-4.adoc (hunk 1 above)
Comment (apinnick) [reply modules/migration-upgrading-migration-tool-4.adoc:9]: I understood that we are not supporting 1.2 upgrade. Who can clarify this?

File: modules/migration-upgrading-migration-tool-4.adoc (hunk 2 above)
Comment (apinnick) [reply modules/migration-upgrading-migration-tool-4.adoc:24]: Are you saying that the migration controller parameters are preserved? In other words, the user has to re-create the migration controller but does not need to reset the parameters in the manifest?

**Comment by XinRedhat** [comment]
https://bz1920404_upgrade-mtc--ocpdocs.netlify.app/openshift-enterprise/latest/migration/migrating_4_2_4/upgrading-migration-tool-4-2-4.html

> 9.Click Workloads → Pods to verify that the ControllerManager, MigrationUI, Restic, and Velero pods are running on the target cluster.
> 10. Click Workloads → Pods to verify that the Restic, and Velero pods are running if you are upgrading MTC on the source cluster.

It seem the description overlap

File: modules/migration-upgrading-migration-tool-4.adoc (hunk 2 above)
Comment (XinRedhat) [reply modules/migration-upgrading-migration-tool-4.adoc:24]: yes, it is on 4.4, 4.7

**Comment by apinnick** [comment]
> https://bz1920404_upgrade-mtc--ocpdocs.netlify.app/openshift-enterprise/latest/migration/migrating_4_2_4/upgrading-migration-tool-4-2-4.html
> 
> > 9.Click Workloads → Pods to verify that the ControllerManager, MigrationUI, Restic, and Velero pods are running on the target cluster.
> > 10. Click Workloads → Pods to verify that the Restic, and Velero pods are running if you are upgrading MTC on the source cluster.
> 
> It seem the description overlap

They do overlap. If someone is upgrading MTC on a 4.x source cluster and a 4.x target cluster, they have to verify that the right pods are running on both clusters. I'll try to find a better way to phrase this.

File: modules/migration-upgrading-migration-tool-4.adoc (hunk 2 above)
Comment (apinnick) [reply modules/migration-upgrading-migration-tool-4.adoc:24]: OK. I'll fix that.

**Comment by eriknelson** [comment]
Why did we add changes related to Automatic operator upgrades for 3 to 4 (https://docs.openshift.com/container-platform/4.6/migration/migrating_3_4/upgrading-migration-tool-3-4.html#migration-upgrading-migration-tool-4_migrating-3-4 does not have anything related to this).

I'm concerned about Automatic operator upgrades because **if your ocp4 MTC automatically upgrades, you aren't likely to notice that happen or it could happen mid-migration, which is likely going to break you because it must be on the same version as your source, and 3.x does not automatically update. It requires manual intervention to keep these in sync.**

**Comment by apinnick** [comment]
> Why did we add changes related to Automatic operator upgrades for 3 to 4 (https://docs.openshift.com/container-platform/4.6/migration/migrating_3_4/upgrading-migration-tool-3-4.html#migration-upgrading-migration-tool-4_migrating-3-4 does not have anything related to this).
> 
> I'm concerned about Automatic operator upgrades because **if your ocp4 MTC automatically upgrades, you aren't likely to notice that happen or it could happen mid-migration, which is likely going to break you because it must be on the same version as your source, and 3.x does not automatically update. It requires manual intervention to keep these in sync.**

Correct. The Operator on the OCP4 cluster should not update automatically or else the versions will not be in sync.

The Operator auto upgrades were part of the single module for upgrading MTC on OCP4 cluster, regardless of the environment. I can add conditions so that auto upgrade is not mentioned in the 3-to-4 migration docs. Maybe I should add an explicit note not to enable subscriptions. I think updates are enabled by default. Will check.

**Comment by eriknelson** [comment]
> > Why did we add changes related to Automatic operator upgrades for 3 to 4 (https://docs.openshift.com/container-platform/4.6/migration/migrating_3_4/upgrading-migration-tool-3-4.html#migration-upgrading-migration-tool-4_migrating-3-4 does not have anything related to this).
> > I'm concerned about Automatic operator upgrades because **if your ocp4 MTC automatically upgrades, you aren't likely to notice that happen or it could happen mid-migration, which is likely going to break you because it must be on the same version as your source, and 3.x does not automatically update. It requires manual intervention to keep these in sync.**
> 
> Correct. The Operator on the OCP4 cluster should not update automatically or else the versions will not be in sync.
> 
> The Operator auto upgrades were part of the single module for upgrading MTC on OCP4 cluster, regardless of the environment. 

AH okay, thank you that explains it. We need to be careful to avoid recommending auto upgrades for the target on ocp3 -> ocp4, it's probably acceptable and even desired on ocp4 -> ocp4 scenarios where both clusters will update.

Maybe it makes sense to explain the issue for ocp3 -> ocp4 case and strongly recommend manual upgrades?

**Comment by XinRedhat** [comment]
LGTM

File: modules/migration-upgrading-migration-tool-4.adoc [hunk 3]
```diff
@@ -4,44 +4,56 @@
...
+[source,yaml]
 ----
-$ oc delete migrationcontroller -n openshift-migration migration-controller
+spec:
+...
+  migration_controller: false
+  migration_ui: false
+  deprecated_cors_configuration: true
 ----
++
```
Comment (bobfuru) [inline modules/migration-upgrading-migration-tool-4.adoc:40]: See previous note about admonition formatting.

**Review by bobfuru — COMMENTED** [review]
A few minor nits and suggestions and otherwise LGTM!

File: modules/migration-installing-cam-operator-ocp-3.adoc [hunk 4]
```diff
@@ -1,14 +1,15 @@
 // Module included in the following assemblies:
 //
 // * migration/migrating_3_4/deploying-cam-3-4.adoc
+
 [id="migration-installing-cam-operator-ocp-3_{context}"]
 ifdef::migrating-3-4[]
-= Installing the {mtc-full} Operator on an {product-title} 3 source cluster
+= Installing the {mtc-full} on an {product-title} 3 source cluster
 
 You can install the {mtc-full} ({mtc-short}) Operator manually on an {product-title} 3 source cluster.
```
[the remark this answers is not in the document]
Comment (apinnick) [reply modules/migration-installing-cam-operator-ocp-3.adoc:9]: Agree. Changed it to "You can install the ... manually on ... cluster." The manual installation requires more than just installing the Operator, so I removed it.

File: modules/migration-installing-cam-operator-ocp-4.adoc [hunk 5]
```diff
@@ -6,30 +6,30 @@
...
+= Installing the {mtc-full} on an {product-title} {product-version} target cluster
 endif::[]
 ifdef::disconnected-3-4,disconnected-target-4-1-4,disconnected-target-4-2-4[]
-= Installing the {mtc-full} Operator on an {product-title} {product-version} target cluster in a restricted environment
+= Installing the {mtc-full} on an {product-title} {product-version} target cluster in a restricted environment
 endif::[]
 
 ifdef::source-4-1-4,source-4-2-4,disconnected-source-4-1-4,disconnected-source-4-2-4[]
-You can install the {mtc-full} ({mtc-short}) Operator on an {product-title} 4 source cluster with the Operator Lifecycle Manager (OLM).
+You can install the {mtc-full} ({mtc-short}) on an {product-title} 4 source cluster by using the Operator Lifecycle Manager (OLM) to install the {mtc-full} Operator.
```
[the remark this answers is not in the document]
Comment (apinnick) [reply modules/migration-installing-cam-operator-ocp-4.adoc:28]: Never heard about this. OK.

File: modules/migration-installing-cam-operator-ocp-4.adoc [hunk 6]
```diff
@@ -51,43 +51,45 @@ ifdef::source-4-1-4[]
 endif::[]
 . Use the *Filter by keyword* field to find the *{mtc-full} Operator*.
 . Select the *{mtc-full} Operator* and click *Install*.
-. On the *Install Operator* page, click *Install*.
+ifdef::migrating-3-4[]
++
+NOTE: Do not change the subscription approval option to *Automatic*. The {mtc-full} Operator version must be the same on the source and the target clusters.
```
[the remark this answers is not in the document]
Comment (apinnick) [reply modules/migration-installing-cam-operator-ocp-4.adoc:56]: OK

File: modules/migration-installing-cam-operator-ocp-4.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bobfuru) [reply modules/migration-installing-cam-operator-ocp-4.adoc]: @adellape should we add this to https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/term_glossary.adoc?

**Pull request merged**